### PR TITLE
R4 profile, XInput backend improvements, etc.

### PR DESCRIPTION
**Pull request by avahe-kellenberger**
Github is currently showing many changes in files which actually have no difference between the master branch and my fork. I'm partially opening this PR to see if it fixes the comparison.

### CODE REVIEW

**Comment by ribbanya** [comment]
@avahe-kellenberger Could you summarize any changes in this PR (or other findings you've come across) that would pertain to #26? I'd like to get that merged at some point.

**Comment by ribbanya** [comment]
By the way, if you use a [two-dot comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#two-dot-git-diff-comparison) then the problem mentioned in the OP should resolve, [like this](https://github.com/JonnyHaystack/HayBox/compare/master..avahe-kellenberger:HayBox:master).

File: HAL/pico/src/comms/XInputBackend.cpp [hunk 1]
```diff
@@ -40,10 +40,12 @@ void XInputBackend::SendReport() {
...
     _report.y = _outputs.y;
-    _report.lb = _outputs.buttonL;
-    _report.rb = _outputs.buttonR;
-    _report.lt = _outputs.triggerLDigital ? 255 : _outputs.triggerLAnalog;
-    _report.rt = _outputs.triggerRDigital ? 255 : _outputs.triggerRAnalog;
+
+    _report.lb = _outputs.triggerLDigital;
+    _report.rb = _outputs.triggerRDigital;
+    _report.lt = _outputs.buttonL ? 255 : _outputs.triggerLAnalog;
+    _report.rt = _outputs.buttonR ? 255 : _outputs.triggerRAnalog;
```
Comment (avahe-kellenberger) [inline HAL/pico/src/comms/XInputBackend.cpp:47]: These seemed to not be working until flipped (lt <-> lb, rt <-> rb).
255 is the max value for when the digital press is registered.

If there are any concerns or issues, let me know. I was testing these settings from yuzu and some steam games.

File: HAL/pico/src/comms/DInputBackend.cpp [hunk 2]
```diff
@@ -33,27 +33,31 @@ void DInputBackend::SendReport() {
...
+    _gamepad->setButton(3, _outputs.y);
+    _gamepad->setButton(4, _outputs.select);
+    _gamepad->setButton(5, _outputs.home);
+    _gamepad->setButton(6, _outputs.start);
+    _gamepad->setButton(7, _outputs.leftStickClick);
+    _gamepad->setButton(8, _outputs.rightStickClick);
+    _gamepad->setButton(9, _outputs.buttonL);
+    _gamepad->setButton(10, _outputs.buttonR);
+    _gamepad->setButton(11, _outputs.triggerLDigital);
+    _gamepad->setButton(12, _outputs.triggerRDigital);
```
Comment (avahe-kellenberger) [inline HAL/pico/src/comms/DInputBackend.cpp:49]: Unnecessary, but helped when using DInput in my game engine. Doesn't really affect anything.

File: HAL/pico/src/comms/DInputBackend.cpp [hunk 3]
```diff
@@ -33,27 +33,31 @@ void DInputBackend::SendReport() {
...
     _gamepad->leftYAxis(255 - _outputs.leftStickY);
     _gamepad->rightXAxis(_outputs.rightStickX);
     _gamepad->rightYAxis(255 - _outputs.rightStickY);
-    _gamepad->triggerLAnalog(_outputs.triggerLAnalog + 1);
-    _gamepad->triggerRAnalog(_outputs.triggerRAnalog + 1);
+
+    // NOTE: This is wrong! ...But works for my needs.
+    // Not sure why L and R aren't working in yuzu with triggerLAnalog and triggerRAnalog.
+    _gamepad->triggerLAnalog(_outputs.triggerLDigital + 1);
+    _gamepad->triggerRAnalog(_outputs.triggerRDigital + 1);
```
Comment (avahe-kellenberger) [inline HAL/pico/src/comms/DInputBackend.cpp:60]: Can revert if needed, reasoning in the comments.

File: HAL/pico/src/comms/NintendoSwitchBackend.cpp [hunk 4]
```diff
@@ -100,43 +100,51 @@ void NintendoSwitchBackend::RegisterDescriptor() {
...
+  _report.b = _outputs.b;
+  _report.a = _outputs.a;
+  _report.x = _outputs.x;
+
+  // TODO: This flips R1/R2 and L1/L2 but the game actually sees b0xx L and R as the triggers.
+  // Maybe the pinout needs to change? Unsure.
+  _report.l = _outputs.triggerLDigital;
+  _report.r = _outputs.triggerRDigital;
+  _report.zl = _outputs.buttonL;
+  _report.zr = _outputs.buttonR;
```
Comment (avahe-kellenberger) [inline HAL/pico/src/comms/NintendoSwitchBackend.cpp:125]: This seems consistent with 3rd party "Pro Controllers", these are always flipped.
Not sure if we could actually fix this in a better way, but I expect not.
Doing this allows for us to reset training mode.

File: HAL/pico/src/comms/NintendoSwitchBackend.cpp [hunk 5]
```diff
@@ -100,43 +100,51 @@ void NintendoSwitchBackend::RegisterDescriptor() {
...
-    TUCompositeHID::_usb_hid.sendReport(_report_id, &_report, sizeof(switch_gamepad_report_t));
+  // Analog outputs
+
+  // Left Stick outputs                             ||     Scaling    ||    || Offset ||
+  _report.lx =        ((_outputs.leftStickX  - 128)   * 1.266  + 128)         + 0.49;     // Rightwards
+  _report.ly = (255 - ((_outputs.leftStickY  - 128)   * 1.256  + 128))        + 1.48;     // Downwards
+
+  // Right Stick outputs
+  _report.rx =        ((_outputs.rightStickX - 128)   * 1.266  + 128)         + 0.49;
+  _report.ry = (255 - ((_outputs.rightStickY - 128)   * 1.256  + 128))        + 1.48;
```
Comment (avahe-kellenberger) [inline HAL/pico/src/comms/NintendoSwitchBackend.cpp:141]: I believe these scaling values were taken from a PR written by @ribbanya but I'm not seeing them looking back.
Maybe ribbanya can comment on this, seems to be working alright but it's not absolutely perfect atm.

File: config/b0xx_r4/config.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  // Create array of input sources to be used.
+  static InputSource *input_sources[] = { gpio_input };
+  size_t input_source_count = sizeof(input_sources) / sizeof(InputSource *);
+
+  ConnectedConsole console = detect_console(pinout.joybus_data);
+
+  /* Select communication backend. */
+  CommunicationBackend *primary_backend;
+  if (console == ConnectedConsole::NONE) {
```
Comment (avahe-kellenberger) [inline config/b0xx_r4/config.cpp:88]: If we want to change buttons holds to fit some uniform standard, that's fine by me.

I do enjoy requiring _no_ button holds for Smash Ultimate on the Switch, though.

File: config/mode_selection.hpp [hunk 7]
```diff
@@ -47,6 +48,7 @@ void select_mode(CommunicationBackend *backend) {
                 )
             );
         } else if (inputs.down) {
+            // TODO: Should I make this switch to UltimateR4?
             set_mode(backend, new Ultimate(socd::SOCD_2IP));
```
Comment (avahe-kellenberger) [inline config/mode_selection.hpp:52]: Decision to be made

File: HAL/pico/src/comms/NintendoSwitchBackend.cpp (hunk 5 above)
Comment (ribbanya) [reply HAL/pico/src/comms/NintendoSwitchBackend.cpp:141]: I honestly don't remember? I just remember the creation of [`Ultimate2`](https://github.com/JonnyHaystack/HayBox/blob/0487bd69b3bdb5230b23da9405d0c354342d8291/src/modes/extra/Ultimate2.cpp) a while ago but that wasn't me.

**Comment by ribbanya** [comment]
That all makes sense! Thanks for taking the time to explain.
